File: group 1/variant11/subscription_service.js
"use strict";

class Request {
    constructor(users, months, nonprofitClaimed, autoRenew) {
        this.users = users;
        this.months = months;
        this.nonprofitClaimed = nonprofitClaimed;
        this.autoRenew = autoRenew;
        Object.freeze(this);
    }
}

class Result {
    constructor(status, total = 0, restorationFee = 0, subscriptionState = null) {
        this.status = status;
        this.total = total;
        this.restorationFee = restorationFee;
        this.subscriptionState = subscriptionState;
        Object.freeze(this);
    }
}

class SubscriptionState {
    currentStatus() {
        throw new Error("Потрібно надати відомості про стан підписки");
    }
}

class NonprofitRegistry {
    isEligible() {
        throw new Error("Потрібно надати підтвердження неприбутковості");
    }
}

class BillingRepository {
    save(request, result) {
        throw new Error("Потрібно надати спосіб фіксації продовження");
    }
}

class SubscriptionService {
    constructor(subscriptions, nonprofits, billing) {
        if (subscriptions == null || nonprofits == null || billing == null) {
            throw new TypeError("Усі залежності мають бути задані");
        }
        this.subscriptions = subscriptions;
        this.nonprofits = nonprofits;
        this.billing = billing;
    }

    process(request) {
        if (!(request instanceof Request)) {
            throw new TypeError("Заявку не задано або її формат некоректний");
        }
        this.checkRange(request.users, 1, 49, "Кількість користувачів");
        this.checkRange(request.months, 1, 12, "Строк продовження");
        if (typeof request.nonprofitClaimed !== "boolean" || typeof request.autoRenew !== "boolean") {
            throw new TypeError("Ознаки пільг мають бути логічними значеннями");
        }
        const state = this.subscriptions.currentStatus();
        if (!["Активна", "Прострочена", "Заблокована"].includes(state)) {
            throw new Error("Невідомий стан підписки");
        }
        if (state === "Заблокована") {
            return new Result("Підписку заблоковано", 0, 0, state);
        }
        let nonprofit = false;
        if (request.nonprofitClaimed) {
            nonprofit = this.nonprofits.isEligible();
            if (typeof nonprofit !== "boolean") {
                throw new Error("Некоректні відомості про неприбутковість");
            }
        }
        let discount;
        if (nonprofit && request.autoRenew) {
            discount = 20;
        } else if (nonprofit) {
            discount = 20;
        } else if (request.autoRenew) {
            discount = 5;
        } else {
            discount = 0;
        }
        const base = request.users * request.months * 20_000;
        let total = base - Math.floor(base * discount / 100);
        if (request.months >= 11) {
            total = Math.max(0, total - 10_000);
        }
        const restorationFee = state === "Прострочена" ? 5_000 : 0;
        const result = new Result("Продовжено", total + restorationFee, restorationFee, "Активна");
        this.billing.save(request, result);
        return result;
    }

    checkRange(value, minimum, maximum, field) {
        if (!Number.isInteger(value) || value < minimum || value > maximum) {
            throw new RangeError("Недопустиме значення: " + field);
        }
    }
}

module.exports = { Request, Result, SubscriptionState, NonprofitRegistry, BillingRepository, SubscriptionService };

File: group 4/variant11/SubscriptionService.java
public final class SubscriptionService {
    public record Request(int users, int months, boolean nonprofitClaimed, boolean autoRenew) {}

    public record Result(String status, long total, long restorationFee, String subscriptionState) {}

    public interface SubscriptionState {
        String currentStatus();
    }

    public interface NonprofitRegistry {
        boolean isEligible();
    }

    public interface BillingRepository {
        void save(Request request, Result result);
    }

    private final SubscriptionState subscriptions;
    private final NonprofitRegistry nonprofits;
    private final BillingRepository billing;

    public SubscriptionService(SubscriptionState subscriptions, NonprofitRegistry nonprofits, BillingRepository billing) {
        this.subscriptions = java.util.Objects.requireNonNull(subscriptions, "Залежність не задано: subscriptions");
        this.nonprofits = java.util.Objects.requireNonNull(nonprofits, "Залежність не задано: nonprofits");
        this.billing = java.util.Objects.requireNonNull(billing, "Залежність не задано: billing");
    }

    public Result process(Request request) {
        if (request == null) {
            throw new IllegalArgumentException("Заявку не задано");
        }
        checkRange(request.users(), 1, 49, "Кількість користувачів");
        checkRange(request.months(), 1, 12, "Строк продовження");
        String state = subscriptions.currentStatus();
        if (!"Активна".equals(state) && !"Прострочена".equals(state) && !"Заблокована".equals(state)) {
            throw new IllegalStateException("Невідомий стан підписки");
        }
        if ("Заблокована".equals(state)) {
            return new Result("Підписку заблоковано", 0, 0, state);
        }
        boolean nonprofit = request.nonprofitClaimed() && nonprofits.isEligible();
        int discount;
        if (nonprofit && request.autoRenew()) {
            discount = 20;
        } else if (nonprofit) {
            discount = 20;
        } else if (request.autoRenew()) {
            discount = 5;
        } else {
            discount = 0;
        }
        long base = request.users() * request.months() * 20_000L;
        long total = base - base * discount / 100;
        if (request.months() >= 11) {
            total = Math.max(0, total - 10_000);
        }
        long restorationFee = "Прострочена".equals(state) ? 5_000 : 0;
        Result result = new Result("Продовжено", total + restorationFee, restorationFee, "Активна");
        billing.save(request, result);
        return result;
    }

    private static void checkRange(int value, int minimum, int maximum, String field) {
        if (value < minimum || value > maximum) {
            throw new IllegalArgumentException("Недопустиме значення: " + field);
        }
    }
}
